index.js

```diff
@@ -51,16 +49,17 @@ async function reportMetric(request, response, startTime, endTime, env) {
 
 export default {
   async fetch(request, env, ctx) {
-    const url = new URL(request.url);
-    const newUrl = `${env.REQUEST_URL}${url.pathname}${url.search}`;
-    // override the request with the new URL
-    const newRequest = new Request(newUrl, request);
-    const reqStartTime = Date.now();
-    const response = await fetch(newRequest);
-    const reqEndTime = Date.now();
+    const cache = caches.default;
+    let response = await cache.match(request);
+    if (!response) {
+      const url = new URL(request.url);
+      const newRequest = new Request(`${env.REQUEST_URL}${url.pathname}${url.search}`, request);
+      response = await fetch(newRequest);
+      ctx.waitUntil(cache.put(request, response.clone()));
```

Comment:
Do we need to manually cache the responses? I believe Cloudflare should be caching `fetch` calls automatically.

See https://github.com/space-meridian/roadmap/issues/206#issuecomment-2532195099

Reply:
> I have removed manual caching and tried running couple of hundred requests to same routes that are currently cached and reported cache status is always `DYNAMIC`.
> 
> I don't think `fetch` automatically caches, but it would if we'd keep our stats API behind the cloudflare cache (like we do now but that'd add additional DNS lookups).

Isn't the cache status going to be always DYNAMIC because the response of the worker execution is never cached?

https://developers.cloudflare.com/cache/concepts/cache-responses/

> A Worker request made a subrequest (`fetch`). In this case, the subrequest will be logged with a cache status, while the main request will be logged with `none/unknown` status (the main request did not hit cache, since Workers sits in front of cache).

Reply:
How good is our understanding of how Cloudflare handles caching when workers call `fetch`? Based on your comments, @pyropy, it seems to me that your view is that a worker calling `fetch` to the origin service (spark-stats on Fly.io) will bypass the Cloudflare cache. From what I read in the docs, and what I would expect a user-friendly solution to implement, my view is that when a worker calls `fetch` with the same request as it's handling, then the request should go through the Cloudflare cache before it reaches the origin service.

How can we validate our understanding?

Here is an idea to consider:

1. Build an endpoint that returns a unique response for each request and configures caching with a relatively long TTL (e.g. 1 hour or more).
2. Invoke that endpoint via the worker you are building. Compare the results returned by the first and the second call.
3. If the worker is implemented correctly, both calls should return the same response.

To keep things simple and also allow us to verify the correctness of our worker in the future too, I propose adding the new endpoint to spark-stats.

For example, in  https://github.com/filecoin-station/spark-stats/blob/main/stats/lib/handler.js:

```diff
   } else if (req.method === 'GET' && url === '/') {
     // health check - required by Grafana datasources
     res.end('OK')
+  } else if (req.method === 'GET' && url === '/now') {
+    res.setHeader('cache-control', `public, max-age=${3_600 /* 1 hour */}`)
+    res.end(new Date().toISOString())
   } else {
     status(res, 404)
   }
```


@juliangruber @pyropy Thoughts?

Reply:
@bajtos Take a look at this [comment](https://github.com/filecoin-station/cf-metrics/pull/1#discussion_r1881800295). 

I also have no expirience with cloudflare workers and did not expect that this kind of behavior is possible (being able to basically intercept the request) to achieve without behaving like a reverse proxy.